Feat: Virtual keyboard support

sudo-gab opened the pull request:
Hello! This PR aims to tackle #292 by implementing a virtual keyboard screen. 

This works by intercepting mouse clicks and checking whether the mouse click occurred inside of a `TextFieldWidget`. 
If it did, the virtual keyboard screen is opened while the listener keeps the reference to the old screen. On virtual keyboard's close, the old screen is re-opened and the clicked TextFieldWidget is injected with the buffer from the virtual keyboard. This may not work in some edge cases like editing a book or a sign, so a unique handler is needed for that (as there is no `TextFieldWidget` in the first place).

### **TODO**

Virtual Keyboard Screen

- [x] Symbol support
- [x] Space support
- [x] Backspace support
- [x] Number support
- [ ] Potential command helpers? (for example, autofill for player names)
- [ ] Styling support? (bold, italics, etc.)

Screen handlers

- [x] `BookEditScreen`
- [ ] `SignEditScreen`

Extra
- [ ] Third party mod compatibility
       ~The main motivation behind this PR was so I could search Roughly Enough Items while playing with a controller. This however, does not seem to work, as clicking on the search text field does not fire a mouse click event. It seems REI captures it - this is something worth investigating.~
         As long as third party mods use an implementation of vanilla Minecraft's TextFieldWidget or ObsidianUI's implementation, it should work. This is dodgy as many mods have different implementations of text fields.
- [x] NeoForge support
- [x] Config value to enable/disable this behavior


This PR is still (early) work in progress. All feedback is welcome.

sudo-gab committed:
Implement basic virtual keyboard support

- Listener for clicks inside of text fields and other text-based screens
- Virtual keyboard screen in a QWERTY layout

Motschen commented:
Great work so far!
Lately, I haven't had the time to add new features to MidnightControls due to the constant need for ports, so this is very appreciated :)

sudo-gab committed:
Move cursor to end of text in book

sudo-gab committed:
Finalize virtual keyboard screen

- Added backspace support
- Added space support
- Added new line support
- Added caps support
- Added symbol support

sudo-gab committed:
Make special key constant private

sudo-gab committed:
Replace KeyInfo with String

- KeyInfo was leftover code from toying around with various implementations. String is sufficient in current implementation.

sudo-gab committed:
Increase displayed lines for newline support

sudo-gab committed:
Add missing row width calculation

sudo-gab committed:
Implement NeoForge support

sudo-gab committed:
Add virtual keyboard config option

sudo-gab commented:
Hey @Motschen,
with [73c5fe1](https://github.com/TeamMidnightDust/MidnightControls/pull/343/commits/73c5fe1a82ecc48c147dbc85bed49b9317751929) I implemented a config option for the virtual keyboard. I'm in a bit of a predicament whether this value should be false or true by default. Also wondering whether there should be a check if the current active controlsMode is equal to Controller before opening the virtual keyboard

sudo-gab committed:
Fix tooltip entry in en_us locale

sudo-gab committed:
Fix handler naming

Motschen commented:
We should probably keep it off by default for the first iteration of this feature. 
When it's more fleshed out, I'd say enabling it by default would not be a bad idea.

As for the controller mode check: Definitely do one. Keyboard /Mouse players won't need a virtual keyboard. 
The situation with Touch Mode is a little different. Perhaps invoking the OS keyboard would work better in this case, as it likely supports multitouch.

sudo-gab committed:
Add obisidianUI text field interop

sudo-gab committed:
Fix virtual keyboard config

Motschen commented:
I'm going to merge this into a feature branch to work on adding localization and styling support.
Feel free to continue working on the other points you listed :)

Motschen committed:
Merge branch 'feat/virtual-keyboard' into feat/virtual-keyboard
